freva-client/src/freva_client/utils/databrowser_utils.py

```diff
@@ -177,3 +189,239 @@ def get_dirs(user: bool = True) -> Path:
+    ) -> Tuple[List[Path], List[xr.Dataset]]:
+        validated_paths: List[Path] = []
+        validated_xarray_datasets: List[xr.Dataset] = []
+
+        for data in user_data:
+            if isinstance(data, str):
+                if any(char in data for char in ['*', '?', '[']):
+                    expanded_paths = self._expand_wildcards(data)
+                    validated_paths.extend(expanded_paths)
+                else:
+                    path = Path(os.path.expanduser(data))
+                    if not path.exists():
+                        logger.warning(f"File path does not exist: {data}")
+                        continue
+                    validated_paths.append(path)
+            elif isinstance(data, xr.Dataset):
+                validated_xarray_datasets.append(data)
+
+        if not validated_paths and not validated_xarray_datasets:
+            raise FileNotFoundError("No valid file paths or xarray datasets found.")
+
+        return validated_paths, validated_xarray_datasets
+
+    def _process_user_data(self,
+                           executor: concurrent.futures.ThreadPoolExecutor) -> None:
+        """Process xarray datasets and file paths using thread pool."""
+        futures = []
+
+        if self.validated_userdata[1]:
+            for batch in self._batch_data(self.validated_userdata[1]):
+                futures.append(
+                    executor.submit(self._process_userdata_in_executor, batch)
+                )
+
+        if self.validated_userdata[0]:
+            for batch in self._batch_data(self.validated_userdata[0]):
+                futures.append(
+                    executor.submit(self._process_userdata_in_executor, batch)
+                )
+
+            for path in self.validated_userdata[0]:
+                if path.is_dir():
+                    for batch in self._batch_data(self._gather_files_from_dir(path)):
+                        futures.append(
+                            executor.submit(self._process_userdata_in_executor, batch)
+                        )
+
+        for future in futures:
+            try:
+                future.result()
+            except Exception as e:  # pragma: no cover
+                logger.error(f"Error processing batch: {e}")
+
+    def _batch_data(
+        self,
+        items: Union[
+            List[Path],
+            Iterator[Path],
+            List[xr.Dataset]
+        ]
+    ) -> Iterator[Union[List[os.PathLike[str]], List[xr.Dataset]]]:
+        """Yield batches of items (either Paths or xarray Datasets)."""
+        batch = []
+        for item in items:
+            if isinstance(item, (Path, xr.Dataset)):
+                batch.append(item)
+                if len(batch) >= self._batch_size:
+                    yield cast(Union[List[os.PathLike[str]], List[xr.Dataset]], batch)
+                    batch = []
+
+        if batch:
+            yield cast(Union[List[os.PathLike[str]], List[xr.Dataset]], batch)
```

Comment:
Do you really need this _batch_data metho? Isn't the thread pool taking care of it already?